---
title: "Option Pricing with Reinforcement Learning [DRAFT]" 
author: "Peyman Kor"
date: 2025-07-21
description: " A short hands-on tutorial on option pricing with reinforcement learning"
categories: [rl, option] 
tags: [rl] 
image: "sdm_rl.png"
format:
  html:
    toc: true # Enable table of contents for better navigation
    toc-location: right # Position the TOC on the left
    #code-fold: true # Allow code folding for cleaner presentation
    number-sections: true # Number the sections in the output
    self-contained: true # Embed resources like CSSCan you hear my voice?  and images into the HTML
    pdf:
      code-block-width: "80%"
      geometry:
        - margin=0.75in
        - textwidth=6.5in
      code-overflow: wrap
---

## Introduction to Option Pricing

American options are a type of financial derivative that give the holder the right, but not the obligation, to buy or sell an underlying asset at a specified price on or before a certain date. Two main types of American options are:

- **Call Option**: Gives the holder the right to buy the underlying asset at the strike price.
- **Put Option**: Gives the holder the right to sell the underlying asset at the strike price.

An Example on how the Amrican Option work can be useful. 

### Example: American Put Option

Imagine you have a put option on a stock with the following details:

- **Option Type**: American Put Option
- **Strike Price (K)**: $100
- **Underlying Asset Price (S)**: $95 initially
- **Expiration Date**: 3 days from now
- **Premium (Initial Price of Option)**: $5

The above info is all the information you have about the option. Though, you do not know how the price of underlying asset (can be simple AAPLE stock) will change over the next 3 days. 

To make example workable, let's assume the price of the underlying asset will change as follows:

- **Price of Underlying Asset over the next 3 days**:

   | Day | Asset Price (S) |
   |-----|-----------------|
   |  0  |      $95        |
   |  1  |      $92        |
   |  2  |      $90        |
   |  3  |      $97        |
 

The holder can choose to exercise the option at any time before or on the expiration date.

For example, lets assusme the holder of the option exercises the option on day 2. This means the holder can sell the underlying asset for $100, (buy back the stock at $90):

The profit of the holder is then:

$$
Profit = (K - S) + Premium
$$

$$
Profit = (100 - 92) - 5 = 3
$$



The holder can choose to exercise the option at any time before or on the expiration date. Though, the question is when is the optimal time to exercise the option? And of course, you cna not know the price of the underlying asset in the future, at day 0.



 <div align="center">
 
 > **💡**  
 > The <b>Option Price</b> (or premium) is the maximum amount a holder is willing to pay to obtain the right to sell the underlying asset at the strike price, at any time before or on the expiration date.
 
 </div>








I think one visualiziation that can help to understand the option price is the following:




![Numbers of Days to 1M Users](option_exp.png){width=100% fig-alt="Chart showing number of days different platforms took to reach 1M users"}

As you can see from the chart below, on the right side of the chart (Put Option) the "Buyer" has profit if the price of the underlying asset is below the strike price (10 \$). Though if the price of underlying asset is above the strike price, the "Buyer" will lose money (equal to the premium paid (1\$))


## American Option vs European Option

Before going to Amarican Option, it is important to understand the difference between American Option and European Option. 

In American Option, the holder can exercise the option at **any time** before the expiration date. Though, in European Option, the holder can only exercise the option at the expiration date.

Essentially, the flexibility that you have in American Option is the one you do not have in European Option. This flexibility comes with costs. American Option is more valuable than European Option.


The difference in the value of American Option and European Option can be attributed to having flexibility in American Option.

## American Option Pricing As Sequential Decision Making

American Option can be viewed as a sequential decision making problem. The holder of the option can choose to exercise the option at any time before the expiration date. The holder can also choose to hold the option until the expiration date.

Let's say you have American Option that expires in 10 days. At each days at days 1 to 10, you need to decide whether to sell (exercise) the option or hold (continue to hold). I tried to visualize this in the following figure:


![Numbers of Days to 1M Users](sdm_rl.png){width=100% fig-alt="Chart showing number of days different platforms took to reach 1M users"}

Now, we cna think about what is the "profit" we get for each alternative we have:

- Alternative I: If we exercise the option, the profit is known. It is $max(0, S_t - K)$, where $S_t$ is the stock price at time $t$ and $K$ is the strike price.
- Alternative II: If we hold the option, the profit is unknown. Therefore, we need find estimate of what will be the **Expected Value** of holding the option at the time.

If we know the Expected Value of holding the option at the time, we can just simply compare the two alternatives and choose the one higher. Simply:

- If $C(S_t, t) > max(0, S_t - K)$, we should exercise the option.

- If $C(S_t, t) < max(0, S_t - K)$, we should hold the option.

here, the $C(S_t, t)$ is the function that gives the idea about the Expected Value of holding the option at the time. It is a function of the stock price and the time.

Now, we need to find the function $C(S_t, t)$.

In summary, the option price of the American Option can be divided into three steps:

1. Find the function $C(S_t, t)$
2. Compare $C(S_t, t)$ with $max(0, S_t - K)$ to decide when to stop and exercise the option.
3. Repeat the process for different paths (to account for underlying uncertainty) and find the average profit.

In the below schematic, I tried to visualize the process:

![Numbers of Days to 1M Users](schematic.png){width=100%}

## Reinforcement Learning

Reinforcement Learning is one learning paradigm that can be used to solve the problem of option pricing.The main idea in Reinforcement Learning is to learn the optimal policy by interacting with the environment.

Think about a child that learns how to play football. The child first will try how shoot (imitating the others), then receive feedback (whether the shoot was goal or not), over time, through this interaction she learn to "reinforce" actions that generates higher reward.

In Reinforcement Learning, we have a similar process. We have an agent that interacts with the environment, and receives feedback (reward) based on the actions it takes. Over time, the agent learn to take actions that generates higher reward.

In the case of option pricing, the agent is the option holder, the environment is the stock market, and the reward is the profit. The agent needs to learn to take actions that generates higher profit.


![Reinforcement Learning](rlimage.png){width=100%}

As you cna see the process of how Reinforcement Learning work is as follows:

1. The agent takes an action.
2. The agent receives feedback (reward) based on the action it takes.
3. The agent moves to new state, and need to take another action.

### Five Elements of Reinforcement Learning

To work with any Reinforcement Learning problem, we need to define the following five elements:

1. **Environment**: The environment is the world that the agent interacts with.
2. **Agent**: The agent is the entity that takes actions and learns from the environment.
3. **Action**: The action is the thing that the agent can do.
4. **State**: The state is the information that the agent has about the environment.
5. **Reward**: The reward is the feedback that the agent receives from the environment.



## Reinforcement Learning for American Option Pricing

As we discussed in the previous section, American option pricing is a **sequential decision-making problem**: at each time step, the holder decides whether to exercise the option or continue holding it. This fits naturally into a **Reinforcement Learning (RL)** framework.

Here we need to translate the problem into a RL framework. The below is summary of the RL elements for option pricing.

### RL Elements for Option Pricing

- **State \(s\)**: Current time and asset price, \(s = (t, x_t)\)
- **Action \(a\)**:  
  - Exercise   
  - Continue 
- **Reward \(r\)**:  
  - If continuing: \(r = 0\)  
  - If exercising: \(r = \max(K - x_t, 0)\) for a put option (or \(\max(x_t - K, 0)\) for a call)
- **State Transition**: Asset price evolves stochastically (e.g., geometric Brownian motion)

### The Q-Value Function

We follow simple Q learning approach as learning method in RL. In Q learning method, we learn the function $Q(s, a)$ that **expected value of the action $a$ in state $s$**. Note that for the action $a=exercise$, the expected value is the immediate exercise value $g(s)$. The main challenge is to estimate the expected value of the action $a=continue$. Though due to continous nature of the underlying asset price, we build a function approximator to estimate the expected value of the action $a=continue$. Therefore, the Q function is written as $Q(s, a; w)$ that is a function of the state $s$ and the action $a$ and the weights $w$ of the function approximator.

$$
Q(s, a; w) = 
\begin{cases}
\phi(s)^T w & \text{if } a = \text{continue} \\
g(s) & \text{if } a = \text{exercise}
\end{cases}
$$

- $\phi(s)$: Feature vector 
- $w$: Weight vector (learned by RL)
- $g(s)$: Immediate exercise value (for the action $a=exercise$)
  
** Feature Vector:**
The feature vector $\phi(s)$ is a vector of features that are used to represent the state $s$. In this case, we use the Laguerre polynomials of normalized price and time.


**Feature Functions in LaTeX:**

The feature vector $\phi(s)$ is constructed using Laguerre-like basis functions of the normalized asset price $\hat{x}_t = x_t / K$, where $K$ is the option's strike price. The same set of functions is also applied to normalized time $\hat{t} = t / T$, where $T$ is the expiry date.

The basis functions for the normalized price are:

$$
\begin{align*}
\phi_0(s) &= 1 \\
\phi_1(s) &= \exp\left(-\hat{x}_t / 2\right) \\
\phi_2(s) &= \exp\left(-\hat{x}_t / 2\right) \left(1 - \hat{x}_t\right) \\
\phi_3(s) &= \exp\left(-\hat{x}_t / 2\right) \left(1 - 2\hat{x}_t + \frac{1}{2}\hat{x}_t^2\right) \\
\phi_4(s) &= \exp\left(-\hat{x}_t / 2\right) \left(1 - 3\hat{x}_t + \frac{3}{2}\hat{x}_t^2\right) \\
\phi_5(s) &= \exp\left(-\hat{x}_t / 2\right) \left(1 - 4\hat{x}_t + 3\hat{x}_t^2 - \frac{1}{6}\hat{x}_t^3\right)
\end{align*}
$$

where $\hat{x}_t = x_t / K$. The same set of functions is applied to normalized time, replacing $\hat{x}_t$ with $\hat{t} = t / T$. In total, there are 12 feature functions (6 for price, 6 for time), and the weight vector $w$ has 12 elements.


### Learning the Optimal Policy

The RL algorithm (e.g., Least Squares Policy Iteration, LSPI) finds the weights \(w^*\) that minimize the difference between predicted and actual rewards over many simulated experiences:

$$
L(w) = \sum_{i=1}^n \left[ Q(s_i, a_i; w) - (r_i + \gamma Q(s'_i, \pi(s'_i); w)) \right]^2
$$

- \(s_i, a_i, r_i, s'_i\): Experience tuple (state, action, reward, next state)
- \(\gamma\): Discount factor
- \(\pi(s'_i)\): Greedy policy (choose action with highest value at next state)

Taking the semi-gradient of the objective function (Equation (18)) gives

$$
\sum_i \phi(s_i) \cdot \left( \phi(s_i)^\top w^* - \lambda \cdot Q\left(s_i', \pi_D(s_i'); w^*\right) \right) = 0
\tag{19}
$$

The term $\pi_D(s_i')$ is a policy operator, where we need to apply the greedy policy. In the case of American options (AO), it is the $\max$ operator between the continuation and exercise values.



This can be written in

$$
A \cdot w^* = b
$$
with
$$
A = \sum_i \phi(s_i) \left( \phi(s_i)^\top - IC_1 \cdot \gamma \cdot \phi(s_i')^\top \right)
$$
and
$$
b = \gamma \sum_i IC_2 \cdot \phi(s_i) \cdot g(s_i')
$$

The term \textit{atomic experience} refers to each data point in the training data of RL, which consists of a stream of $(s_i, s_i')$. The $m \times m$ matrix $A$ and $m \times 1$ (where $m$ is the size of the feature function vector, here $m=12$) matrix $b$ are accumulated at each atomic experience $(s_i, s_i')$:
$$
A \leftarrow A + \phi(s_i) \left( \phi(s_i)^\top - IC_1 \cdot \gamma \cdot \phi(s_i')^\top \right)
$$
$$
b \leftarrow b + \gamma \cdot IC_2 \cdot \phi(s_i) \cdot g(s_i')
$$

Thus, $w^*$ is solved as
$$
w^* = A^{-1} b
$$
and updates the Q-value function approximation $Q(s, a; w^*)$ [23]. The update of the Q-value function reflects the update in the “decision policy” during the training phase of the Reinforcement Learning method. Unlike classical RL methods [8], which update weights after each individual experience $(s = x_t, s' = x_{t+1})$, a batch (set of experiences) training approach is implemented in this work.


## Python Implementation

- $T$ and $S$ are time points:
$$log(x_T) \sim \mathcal{N}(log(x_S) + (\mu-\frac{\sigma^2}{2})(T-S), \sigma^2(T-S))$$

- Where:
    - $x_T$ is the price of the asset at time $T$
    - $\mu$ is the drift
    - $\sigma$ is the volatility
- We can rewrite equation so that it is in terms of $x_T$:

$$x_T \sim  \exp(\mathcal{N}(log(x_S) + (\mu-\frac{\sigma^2}{2})(T-S), \sigma^2(T-S)))$$

In [1]:
from typing import List, Tuple
import numpy as np

def simulate_GBM_training(
    expiry_time: float,
    num_intervals: int,
    num_simulations: int,
    spot_price: float,
    interest_rate: float,
    volatility: float
) -> List[Tuple[int, float, float]]:

    """
    Generates simulated stock price data for training purposes.

    Parameters:
        expiry_time (float): expiration date (year).
        num_intervals (int): Number of exercise time steps.
        num_simulations (int): Number of independent price paths to generate.
        spot_price (float): Starting stock price, spot price
        interest_rate (float): Risk-free interest rate.
        volatility (float): Volatility of the stock price.

    Returns:
        List[Tuple[int, float, float]]: List of tuples containing (time step, current price, next price).
    """

    results = []
    time_step_size = expiry_time / num_intervals
    volatility_squared = volatility ** 2

    for _ in range(num_simulations):

        current_price = spot_price

        for step in range(num_intervals):
            mean = np.log(current_price) + (interest_rate - 0.5 * volatility_squared) * time_step_size
            variance = volatility_squared * time_step_size
            next_price = np.exp(np.random.normal(mean, np.sqrt(variance)))
            results.append((step, current_price, next_price))
            current_price = next_price

    return results

### Initial Data 

In [2]:
# data for training

expiry_time_val = 1
num_intervals_val = 50
num_simulations_val = 10000
spot_price_val = 36
strike_val = 40
interest_rate_val = 0.06
volatility_val = 0.2

In [3]:
RL_data_training = simulate_GBM_training(
    expiry_time=expiry_time_val,
    num_intervals=num_intervals_val,
    num_simulations=num_simulations_val,
    spot_price=spot_price_val,
    interest_rate=interest_rate_val,
    volatility=volatility_val
)

# Example of the first 3 data points
print(RL_data_training[:3])

[(0, 36, 36.25112364343208), (1, 36.25112364343208, 35.798687306235884), (2, 35.798687306235884, 35.57008138835434)]


The output of `print(RL_data_training[:3])` here shows the first three simulated steps from the generated Geometric Brownian Motion (GBM) training data.

If you look closer, each element in `RL_data_training` is a tuple of the form (step, current_price, next_price):
  - step: the time step index within a simulation path (from 0 up to num_intervals-1)
  - current_price: the simulated asset price at the time step $t$
  - next_price: the simulated asset price at the time step $t+1$, after applying the GBM update

Since there are multiple simulation paths, the data is a flat list of all steps from all simulations.

For example, the output might look like:
  [(0, 36, 36.12), (1, 36.12, 35.98), (2, 35.98, 36.25)]
This means:
  - At step 0, the price moved from 36 to 36.12
  - At step 1, the price moved from 36.12 to 35.98
  - At step 2, the price moved from 35.98 to 36.25



### Defining Feature Functions

The feature vector $\phi(s)$ is constructed using Laguerre-like basis functions of the normalized asset price $\hat{x}_t = x_t / K$, where $K$ is the option's strike price. The same set of functions is also applied to normalized time $\hat{t} = t / T$, where $T$ is the expiry date.

The basis functions for the normalized price are:

$$\phi_0(s) = 1$$

$$\phi_1(s) = \exp\left(-\hat{x}_t / 2\right)$$

$$\phi_2(s) = \exp\left(-\hat{x}_t / 2\right) \left(1 - \hat{x}_t\right)$$

$$\phi_3(s) = \exp\left(-\hat{x}_t / 2\right) \left(1 - 2\hat{x}_t + \frac{1}{2}\hat{x}_t^2\right)$$

$$\phi_4(s) = \exp\left(-\hat{x}_t / 2\right) \left(1 - 3\hat{x}_t + \frac{3}{2}\hat{x}_t^2\right)$$

$$\phi_5(s) = \exp\left(-\hat{x}_t / 2\right) \left(1 - 4\hat{x}_t + 3\hat{x}_t^2 - \frac{1}{6}\hat{x}_t^3\right)$$

In [14]:
from numpy.polynomial.laguerre import lagval
from typing import List, Tuple, Callable

def make_feature_functions(t_s: Tuple[float, float],
        strike: float, expiry: float, num_laguerre:int=6) -> List[Callable[[Tuple[float, float]],
                                                                                             float]]:

    ident = np.eye(num_laguerre)

    def time_feature(t_s, i, expiry, ident):
        return np.exp(-t_s[0] / (2 * expiry)) * lagval(t_s[0] / expiry, ident[i])

    def price_feature(t_s, i, strike, ident):
        return np.exp(-t_s[1] / (2 * strike)) * lagval(t_s[1] / strike, ident[i])

    features: List[Callable[[Tuple[float, float]], float]] = []

    for i in range(num_laguerre):
        features.append(time_feature(t_s, i, expiry, ident))

    for i in range(num_laguerre):
        features.append(price_feature(t_s, i, strike, ident))

    return features

### Finding the optimal weights

The optimal weights $w^*$ are found by solving the linear system of equations.

In [ ]:
import numpy as np

def get_weights(num_laguerre, expiry, num_steps, rate, training_data, strike_price,
                training_iters, epsilon):
    """
    Computes the weights for option pricing model using regression and feature functions.

    Parameters:
        num_laguerre (int): Number of Laguerre polynomials used in feature functions.
        expiry (float): Time to expiry of the option.
        num_steps (int): Number of time steps in the model.
        rate (float): Risk-free interest rate.
        training_data (List[Tuple[int, float, float]]): List of tuples (time step, current price, next price).
        strike (float): Strike price of the option.
        training_iters (int): Number of iterations for regression.
        epsilon (float): Regularization parameter for regression.

    Returns:
        np.ndarray: Calculated weights for the pricing model.
    """
    dt = expiry / num_steps
    gamma = np.exp(-rate * dt)
    num_features = num_laguerre * 2

    states = [(i * dt, s) for i, s, _ in training_data]
    next_states = [((i + 1) * dt, s1) for i, _, s1 in training_data]

    feature_vals = np.array([make_feature_functions(state, strike_price, expiry) for state in states])
    next_feature_vals = np.array([make_feature_functions(next_state, strike_price, expiry) for next_state in next_states])

    non_terminal_flags = np.array([index < num_steps - 1 for index, _, _ in training_data])
    exercise_value = np.array([max(strike_price - stock_price, 0) for _, stock_price in next_states])

    wts = np.ones(num_features)

    for _ in range(training_iters):

        a_inv = np.eye(num_features) / epsilon
        b_vec = np.zeros(num_features)
        cont = np.dot(next_feature_vals, wts)
        cont_cond = non_terminal_flags * (cont > exercise_value)

        for i in range(len(training_data)):
            phi1 = feature_vals[i]
            phi2 = phi1 - cont_cond[i] * gamma * next_feature_vals[i]

            # Sherman-Morrison formula for matrix a_inv
            temp = a_inv.T.dot(phi2)
            a_inv -= np.outer(a_inv.dot(phi1), temp) / (1 + phi1.dot(temp))
            # Sherman-Morrison formula fro matrix a_inv

            # matrix b update according to the formula
            b_vec += phi1 * (1 - cont_cond[i]) * exercise_value[i] * gamma

        wts = a_inv.dot(b_vec)
        #print(b_vec)
        #print(wts)
        #print(a_inv)

    return wts

In [16]:
# Example usage
num_laguerre_val = 6
training_iters_val = 4
epsilon_val = 1e-5
strike_val = 40

weights = get_weights(num_laguerre=num_laguerre_val, expiry=expiry_time_val, num_steps=num_intervals_val,
                        rate=interest_rate_val, training_data=RL_data_training, strike_price=strike_val,
                        training_iters=training_iters_val, epsilon=epsilon_val)

print("Weights: ", weights)

[ 1935.97224426 -1253.09249878   171.72387314   440.71785831
  -315.13029957    65.09910762 -2195.97189808  1598.36993027
  2208.04697418 -7915.28777695  8191.92101288 -3046.98758793]
[   87.57666874   206.81636238  -397.48112965   175.34454441
    75.76797581   -57.44999552  -476.55812836  1249.33361053
  -679.98515701 -2319.04568481  3734.40523529 -1666.68296909]
[  -57.35607052   336.11819983  -420.27219009    62.36166763
   194.37779379   -94.37842882  -290.11853361  1018.45205879
  -770.85533905 -1490.55202293  2738.25713348 -1266.44455338]
[  -19.48214293   299.34066296  -384.54959559    37.16889548
   202.14182019   -94.30528858  -294.45013809   921.03279877
  -638.83655739 -1415.18053627  2506.02262306 -1145.66849947]
Weights:  [  -19.48214293   299.34066296  -384.54959559    37.16889548
   202.14182019   -94.30528858  -294.45013809   921.03279877
  -638.83655739 -1415.18053627  2506.02262306 -1145.66849947]


### Generate the Test Data 

In [17]:
def scoring_sim_data(
    expiry_time: float,
    num_intervals: int,
    num_simulations_test: int,
    spot_price: float,
    interest_rate: float,
    volatility: float
) -> np.ndarray:

    paths: np.ndarray = np.empty([num_simulations_test, num_intervals + 1])
    time_step_size = expiry_time / num_intervals

    #vol2: float = volatility * volatility
    volatility_squared = volatility ** 2

    for i in range(num_simulations_test):
        paths[i, 0] = spot_price
        for step in range(num_intervals):
            mean: float = np.log(paths[i, step]) + (interest_rate - volatility_squared / 2) * time_step_size
            variance = volatility_squared * time_step_size
            paths[i, step + 1] = np.exp(np.random.normal(mean, np.sqrt(variance)))
    return paths[:,1:]

### Finding the Option Price


In [18]:
def scoring_sim_data(
    expiry_time: float,
    num_intervals: int,
    num_simulations_test: int,
    spot_price: float,
    interest_rate: float,
    volatility: float
) -> np.ndarray:

    paths: np.ndarray = np.empty([num_simulations_test, num_intervals + 1])
    time_step_size = expiry_time / num_intervals

    #vol2: float = volatility * volatility
    volatility_squared = volatility ** 2

    for i in range(num_simulations_test):
        paths[i, 0] = spot_price
        for step in range(num_intervals):
            mean: float = np.log(paths[i, step]) + (interest_rate - volatility_squared / 2) * time_step_size
            variance = volatility_squared * time_step_size
            paths[i, step + 1] = np.exp(np.random.normal(mean, np.sqrt(variance)))
    return paths[:,1:]

In [19]:
num_simulations_test = 10000

paths_test = scoring_sim_data(
    expiry_time=expiry_time_val,
    num_intervals=num_intervals_val,
    num_simulations_test=num_simulations_test,
    spot_price=spot_price_val,
    interest_rate=interest_rate_val,
    volatility=volatility_val
)

paths_test[:2]

array([[36.25546759, 36.8459042 , 36.23810201, 36.57563596, 37.05929142,
        37.53855102, 36.05719982, 36.87279366, 39.4033157 , 39.93436394,
        41.95659503, 40.08263779, 40.6096482 , 40.4663088 , 41.40580106,
        42.6960731 , 42.78037411, 43.35851883, 44.2514124 , 43.38252324,
        44.94390416, 43.39866876, 43.94583491, 44.93006643, 44.94871952,
        46.53193819, 44.97876723, 47.59628139, 47.03834805, 46.55573369,
        47.45076548, 47.00464637, 47.44650599, 46.3681192 , 45.84378305,
        45.40798696, 43.51242877, 43.4503    , 43.87416368, 43.06175161,
        42.56575755, 42.43638311, 43.0564607 , 42.86506152, 43.46328668,
        41.24391947, 41.02859932, 40.61937743, 40.70874376, 42.99537259],
       [36.87572603, 37.8677158 , 36.91667468, 37.78780229, 37.79946953,
        37.22470354, 38.05376542, 37.99941835, 39.48314209, 38.3034728 ,
        34.78798265, 33.56007705, 31.63469746, 31.49812295, 31.50675108,
        31.68254114, 32.5331868 , 34.2634487 , 33.

In [20]:
def calculate_option_price(
    stock_paths: np.ndarray,     # This is a 2D array with stock price paths
    option_expiry: float,        # The time until the option expires, in years
    interest_rate: float,        # The annual interest rate, as a decimal
    strike_price: float,
    weights_rl: List           # The price at which the option can be exercised
) -> float:
    # Number of different price paths

    num_paths = stock_paths.shape[0]
    # Number of time steps in each path
    num_time_steps = stock_paths.shape[1]

    # This will hold the final option prices for each path
    option_prices = np.zeros(num_paths)
    # Time interval between steps
    dt = option_expiry / num_time_steps

    # Loop over each stock price path
    for path_index, price_path in enumerate(stock_paths):
        # Process each time step in the path
        time_step = 1
        while time_step <= num_time_steps:
            # Calculate the current time in years
            current_time = time_step * dt
            # Get the stock price at the current step
            current_stock_price = price_path[time_step - 1]
            # Calculate the profit if the option is exercised now
            exercise_value = max(strike_price - current_stock_price, 0)
            # Simulate the feature values for mathematical modeling (using a placeholder function)
            state = (current_time, current_stock_price)
            feature_values = make_feature_functions(state, strike_price, option_expiry)
            # Calculate the theoretical price if waiting (using machine learning or some model)
            continue_price = np.dot(feature_values, weights_rl) if time_step < num_time_steps else 0

            # Increment the time step
            time_step += 1
            # Decide whether to exercise the option now or later
            if (exercise_value >= continue_price) and (exercise_value > 0):
                # Update the price for this path and discount it by the interest rate
                option_prices[path_index] = np.exp(-interest_rate * current_time) * exercise_value
                # Move to the end of the path since we've decided to exercise the option
                time_step = num_time_steps + 1

    # Return the average of the option prices across all paths
    return np.average(option_prices)

In [21]:
option_price = calculate_option_price(
    stock_paths=paths_test,
    option_expiry=expiry_time_val,
    interest_rate=interest_rate_val,
    strike_price=strike_val,
    weights_rl=weights
)
print("Option Price using RL Method:", option_price)

Option Price using RL Method: 4.431395568006979
